pkg/api/customization/globalnamespaceaccess/access_common.go

```diff
@@ -110,66 +108,184 @@ func (ma *MemberAccess) EnsureRoleInTargets(targetProjects, roleTemplates []stri
+					}
+					if rt.ProjectCreatorDefault && rt.Builtin {
+						// this is the "project-owner" role
+						inputRolesContainProjectOwnerRole = true
+					}
+				}
+			}
+			// check if caller is project-owner, project-member or cluster-owner
+			if callerIsProjectOwner && !customRolesFound {
+				// project-owner should be allowed to add any built-in project roles
+				continue
+			}
+			if callerIsProjectMember && !customRolesFound && !inputRolesContainProjectOwnerRole {
+				// project-member should be allowed to add any built-in project role, EXCEPT the project-owner role
+				continue
+			}
+
+			// check if caller is cluster-owner
+			crtbs, err := ma.CrtbLister.List(cname, labels.NewSelector())
+			if err != nil {
+				return err
+			}
+			for _, crtb := range crtbs {
+				if crtb.UserName == callerID {
+					rt, err := ma.RoleTemplateLister.Get("", crtb.RoleTemplateName)
+					if err != nil {
+						return err
+					}
+					if rt.ClusterCreatorDefault && rt.Builtin {
+						// caller is the owner of the cluster that this project belongs to, no need to check other crtbs
+						callerIsClusterOwner = true
+						break
+					}
+				}
+			}
+			if callerIsClusterOwner && !customRolesFound {
+				// cluster-owner should be allowed to add any built-in roles
+				continue
+			}
+			// either the user is not one of these: project-owner, project-member or cluster-owner, OR
+			// the passed in roles have some custom roles which the user does not have prtbs/crtbs for
+			p, err := ma.ProjectLister.Get(cname, pname)
+			if err != nil {
+				return err
+			}
+			projectName := pname
+			if p.Spec.DisplayName != "" {
+				projectName = p.Spec.DisplayName
+			}
+			errMsg := fmt.Sprintf("user %v does not have following roles in project %v: %v", callerID, projectName, rolesNotFound)
 			return httperror.NewAPIError(httperror.PermissionDenied, errMsg)
 		}
 	}
+
+	if len(newClusterRoleTemplateMap) == 0 {
+		return nil
+	}
 	for cname := range clusters {
 		clusterRoleTemplateFoundCount := 0
 		clusterRoleTemplateFoundMap := make(map[string]bool)
+		clusterOwner := false
 		crtbs, err := ma.CrtbLister.List(cname, labels.NewSelector())
 		if err != nil {
 			return err
 		}
 		for _, crtb := range crtbs {
-			if crtb.UserName == callerID && newClusterRoleTemplateMap[crtb.RoleTemplateName] {
-				clusterRoleTemplateFoundMap[crtb.RoleTemplateName] = true
-				clusterRoleTemplateFoundCount++
+			if crtb.UserName == callerID {
+				if _, ok := newClusterRoleTemplateMap[crtb.RoleTemplateName]; ok {
+					clusterRoleTemplateFoundMap[crtb.RoleTemplateName] = true
+					clusterRoleTemplateFoundCount++
+				}
+				if clusterOwner {
+					// we already found a crtb with roletemplate cluster-owner for the caller in this cluster
+					continue
+				}
+				rt, err := ma.RoleTemplateLister.Get("", crtb.RoleTemplateName)
+				if err != nil {
+					return err
+				}
+				if rt.ClusterCreatorDefault && rt.Builtin {
+					clusterOwner = true
+				}
```

Comment:
Maybe we can avoid this loop in cases where we already found that user is clusterOwner for some clusters  from line#196? We can store if clusterOwnerMap[string]bool, set it true for those clusters that we already found this info for.

Reply:
@prachidamle I'm not sure I follow, we can save the clusters that the user is owner of, and avoid getting the roletemplate in this loop, but this loop is still need to check for other cluster roles

Reply:
yeah @mrajashree this code is still needed, but we can first check if clusterOwnerMap[cname] is true - if yes then no need to do the check for ```if rt.ClusterCreatorDefault && rt.Builtin {```?